### PR TITLE
fix(config): Force enabledManagers option

### LINKED ISSUE #9408: enabledManagers option should enable what is disabled by default
**What Renovate type, platform and version are you using?**

Doesn't matter.

**Describe the bug**

```json
{
  "enabledManagers": ["gradle-lite"],
}
```

should work as an equivalent to 

```json
{
  "gradle": {
    "enabled": false
  },
  "gradle-lite": {
    "enabled": true
  }
}
```

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

Massage config by converting `enabledManagers` items to corresponding `enabled: true | false` flags for each manager.

## Context:

Closes #9408

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Example: https://github.com/renovate-testing/test-9408-enabled-managers

### CODE REVIEW

File: lib/config/enabled-managers.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  if (is.nonEmptyArray(enabledManagers)) {
+    logger.debug({ enabledManagers }, 'Applying enabled managers filtering');
+    const enabledSet = new Set([...enabledManagers]);
+    for (const managerName of getManagerList()) {
+      const manager = config[managerName] as ManagerConfig;
+      if (manager) {
+        manager.enabled = enabledSet.has(managerName);
+      }
+    }
+    enabledManagers.splice(0, enabledSet.size);
```
Comment (rarkins) [inline lib/config/enabled-managers.ts:19]: What's the purpose of this line?

File: lib/workers/repository/extract/index.ts [hunk 2]
```diff
@@ -14,39 +13,36 @@ import { getManagerPackageFiles } from './manager-files';
...
-          extractList.push(regexManagerConfig);
-        }
+        tryConfig(mergeChildConfig(managerConfig, regexManager));
       }
     } else {
-      managerConfig.fileList = getMatchingFiles(managerConfig, fileList);
-      if (managerConfig.fileList.length) {
-        extractList.push(managerConfig);
-      }
+      tryConfig(managerConfig);
```
Comment (rarkins) [inline lib/workers/repository/extract/index.ts:42]: Is this file a pure refactor or is it necessary to force enabledManagers?

File: lib/config/enabled-managers.ts (hunk 1 above)
Comment (zharinov) [reply lib/config/enabled-managers.ts:19]: Reset the array, actually it's better to not do this

File: lib/workers/repository/extract/index.ts (hunk 2 above)
Comment (zharinov) [reply lib/workers/repository/extract/index.ts:42]: It's a pure refactor for avoiding similar condition guards for both if-else branches

**Comment by zharinov** [comment]
BTW, this change can be breaking for those who relied on old (undocumented) behavior

File: lib/workers/repository/extract/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/workers/repository/extract/index.ts:42]: Is it possible to split the refactor first?

File: lib/config/enabled-managers.spec.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { applyEnabledManagersFilter } from './enabled-managers';
+import { RenovateConfig } from './types';
+
+describe('config/enabled-managers', () => {
```
Comment (viceice) [inline lib/config/enabled-managers.spec.ts:4]:
```suggestion
describe(getName(__filename), () => {
```

File: lib/config/enabled-managers.spec.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      enabled: false,
+    },
+  };
+  it('ignores config changes when enabledManagers is empty', () => {
+    expect(applyEnabledManagersFilter(config)).toStrictEqual(config);
+    expect(
+      applyEnabledManagersFilter({ ...config, enabledManagers: ['foobar'] })
+    ).toStrictEqual({ ...config, enabledManagers: ['foobar'] });
+  });
+  it('changes enabled flag for enabled managers', () => {
```
Comment (viceice) [inline lib/config/enabled-managers.spec.ts:19]:
```suggestion

  it('changes enabled flag for enabled managers', () => {
```

File: lib/config/enabled-managers.spec.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+describe('config/enabled-managers', () => {
+  const config: RenovateConfig = {
+    npm: {
+      enabled: true,
+    },
+    maven: {
+      enabled: false,
+    },
+  };
+  it('ignores config changes when enabledManagers is empty', () => {
```
Comment (viceice) [inline lib/config/enabled-managers.spec.ts:13]:
```suggestion

  it('ignores config changes when enabledManagers is empty', () => {
```

Readability

File: lib/config/enabled-managers.spec.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { applyEnabledManagersFilter } from './enabled-managers';
+import { RenovateConfig } from './types';
```
Comment (viceice) [inline lib/config/enabled-managers.spec.ts:2]:
```suggestion
import type { RenovateConfig } from './types';
```

File: lib/config/enabled-managers.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+): RenovateConfig {
+  const { enabledManagers = [] } = config;
+  if (is.nonEmptyArray(enabledManagers)) {
+    logger.debug({ enabledManagers }, 'Applying enabled managers filtering');
+    const enabledSet = new Set([...enabledManagers]);
+    for (const managerName of getManagerList()) {
+      const manager = config[managerName] as ManagerConfig;
+      if (manager) {
+        manager.enabled = enabledSet.has(managerName);
+      }
```
Comment (viceice) [inline lib/config/enabled-managers.ts:17]: shouldn't it disable all other managers? Are you sure the manager config is always there. We need to be sure it disables all other manager even there is no manager config yet.

File: lib/workers/repository/init/index.ts [hunk 8]
```diff
@@ -21,6 +22,7 @@ export async function initRepo(
   config = await initApis(config);
   config = await getRepoConfig(config);
   checkIfConfigured(config);
+  config = applyEnabledManagersFilter(config);
```
Comment (zharinov) [inline lib/workers/repository/init/index.ts:25]: > shouldn't it disable all other managers? Are you sure the manager config is always there. We need to be sure it disables all other manager even there is no manager config yet.

@viceice It seemed to me that all relevant data is ready at this point, isn't it?

Comment (viceice) [reply lib/workers/repository/init/index.ts:25]: I'm not sure 🤷‍♂️

**Comment by viceice** [comment]
Maybe better target this to v25?

**Comment by HonkingGoose** [comment]
> BTW, this change can potentially be breaking for those who relied on old (undocumented) behavior. Though, I can't imagine any example of its _useful_ abuse.

If the behavior was always there, and we're now changing it, I would consider that a "breaking" change. Even if the workflow we're "breaking" is really stupid. 😄 

workflow

From XKCD 1172 "Workflow".

---

On topic: Should we label this PR with the `breaking` tag then?

**Comment by zharinov** [comment]
Yep, that's exactly my concern

**Comment by rarkins** [comment]
Yes, wow should merge it into v25 as breaking. I also worry it might be *broken*

**Comment by rarkins** [comment]
I would prefer to skip this complexity. We don't often have disabled by default managers

**Pull request closed without merging**